src/featuremap/ugbio_featuremap/somatic_featuremap_utils.py

```diff
@@ -0,0 +1,155 @@
+import os
+import subprocess
+import tempfile
+from os.path import join as pjoin
+
+import pandas as pd
+import pysam
+from ugbio_core.logger import logger
+
+
+def _create_variant_bed(merged_vcf, bed_file):
+    """Create a BED file from VCF variants."""
+    cmd_bcftools = ["bcftools", "query", "-f", "%CHROM\t%POS\t%END\n", merged_vcf]
+    cmd_awk = ["awk", 'BEGIN{OFS="\t"} {print $1, $2, $3+1}']
+    with open(bed_file, "w") as out_file:
+        p1 = subprocess.Popen(cmd_bcftools, stdout=subprocess.PIPE)
+        p2 = subprocess.Popen(cmd_awk, stdin=p1.stdout, stdout=out_file)
+        p1.stdout.close()
+        p2.communicate()
+        p1.wait()
+
+
+def _find_closest_tandem_repeats(bed1, bed2, output_file):
+    """Find closest tandem repeats for each variant."""
+    cmd_bedtools = ["bedtools", "closest", "-D", "ref", "-a", bed1, "-b", bed2]
+    cmd_cut = ["cut", "-f1-2,5-8"]
+    with open(output_file, "w") as out_file:
+        p1 = subprocess.Popen(cmd_bedtools, stdout=subprocess.PIPE)
+        p2 = subprocess.Popen(cmd_cut, stdin=p1.stdout, stdout=out_file)
+        p1.stdout.close()
+        p2.communicate()
+        p1.wait()
+
+
+def _process_tandem_repeat_data(tr_tsv_file):
+    """Process tandem repeat data and calculate metrics."""
+    df_tr_info = pd.read_csv(tr_tsv_file, sep="\t", header=None)
+    df_tr_info.columns = ["chrom", "pos", "TR_START", "TR_END", "TR_SEQ", "TR_DISTANCE"]
+    df_tr_info["TR_LENGTH"] = df_tr_info["TR_END"] - df_tr_info["TR_START"]
+
+    extracted = df_tr_info["TR_SEQ"].str.extract(r"\((\w+)\)")
+    df_tr_info["TR_SEQ_UNIT_LENGTH"] = extracted[0].str.len()
```

Comment:
If you want the length, why not add a column to the input tandem repeat bed file instead of doing that afterwards per line in pandas?

Same for TR_LENGTH above

I think it could make this entire pandas segment unnecessary

Reply:
Alternatively this can also happen in the downstream code where you anyway manipulate the features.